app/src/main/java/jaelson/wagner/silva/upcomingmovies/ui/movies/upcoming/UpcomingMoviesFragment.java

```diff
@@ -0,0 +1,160 @@
+package jaelson.wagner.silva.upcomingmovies.ui.movies.upcoming;
+
+import android.content.Intent;
+import android.content.res.Configuration;
+import android.os.Bundle;
+import android.support.annotation.Nullable;
+import android.support.v4.app.Fragment;
+import android.support.v4.widget.SwipeRefreshLayout;
+import android.support.v7.widget.GridLayoutManager;
+import android.support.v7.widget.RecyclerView;
+import android.view.LayoutInflater;
+import android.view.View;
+import android.view.ViewGroup;
+import jaelson.wagner.silva.upcomingmovies.R;
+import jaelson.wagner.silva.upcomingmovies.UpcomingMoviesApp;
+import jaelson.wagner.silva.upcomingmovies.data.model.movies.Movie;
+import jaelson.wagner.silva.upcomingmovies.di.module.UpcomingMoviesModule;
+import jaelson.wagner.silva.upcomingmovies.ui.common.EndlessRecyclerViewScrollListener;
+import jaelson.wagner.silva.upcomingmovies.ui.movies.details.MovieDetailsActivity;
+import timber.log.Timber;
+
+import javax.inject.Inject;
+import java.util.List;
+
+/**
+ * Created by jaelsonwagner@gmail.com.
+ */
+
+public class UpcomingMoviesFragment
+    extends Fragment
+    implements UpcomingMoviesContract.View {
+
+  private static final int FIRST_PAGE = 1;
+  private static final int SPAN_COUNT_2 = 2;
+  private static final int SPAN_COUNT_3 = 3;
+  private static final int NUMBER_OF_REMAINING_ITEMS = 4;
+  @Inject
+  UpcomingMoviesContract.Presenter presenter;
+  private RecyclerView recyclerView;
+  private EndlessRecyclerViewScrollListener endlessRecyclerViewScrollListener;
+  private int nextPageNumber = FIRST_PAGE;
+  private int previousPageNumber;
+  private UpcomingMoviesAdapter upcomingMoviesAdapter;
+  private SwipeRefreshLayout swipeRefreshLayout;
+
+  public static UpcomingMoviesFragment newInstance() {
+    return new UpcomingMoviesFragment();
+  }
+
+  @Nullable
+  @Override
+  public View onCreateView(LayoutInflater inflater,
+                           @Nullable ViewGroup container,
+                           @Nullable Bundle savedInstanceState) {
+    return inflater.inflate(R.layout.fragment_upcoming_movies, container, false);
+  }
+
+  @Override
+  public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
+    super.onViewCreated(view, savedInstanceState);
+
+    UpcomingMoviesApp.getApplicationComponent()
+        .plus(new UpcomingMoviesModule(this))
+        .inject(this);
+
+    setupUI();
+
+    presenter.fetchUpcomingMovies(FIRST_PAGE, previousPageNumber);
```

Comment:
How would you avoid an extra request when moving back to this page or on orientation change?

Reply:
Well, until now, all apps I have worked don't support change orientation. So, I have just theoretical knowledge. But, to solve that, I would use Bundle (savedInstanceState), to handle that. Sorry to delivery that way. Given my deadline and my lack of time, there is some technical debts.